lib/elixir/test/elixir/code_normalizer/general_test.exs

```diff
@@ -0,0 +1,616 @@
+    end
+
+    test "string" do
+      assert quoted_to_string(quote(do: "")) == ~S/""/
+      assert quoted_to_string(quote(do: "abc")) == ~S/"abc"/
+      assert quoted_to_string(quote(do: "#{"abc"}")) == ~S/"#{"abc"}"/
+    end
+
+    test "last arg keyword list" do
+      assert quoted_to_string(quote(do: foo([]))) == "foo([])"
+      assert quoted_to_string(quote(do: foo(x: y))) == "foo(x: y)"
+      assert quoted_to_string(quote(do: foo(x: 1 + 2))) == "foo(x: 1 + 2)"
+      assert quoted_to_string(quote(do: foo(x: y, p: q))) == "foo(x: y, p: q)"
+      assert quoted_to_string(quote(do: foo(a, x: y, p: q))) == "foo(a, x: y, p: q)"
+
+      assert quoted_to_string(quote(do: {[]})) == "{[]}"
+      assert quoted_to_string(quote(do: {[a: b]})) == "{[a: b]}"
+      assert quoted_to_string(quote(do: {x, a: b})) == "{x, [a: b]}"
+      assert quoted_to_string(quote(do: foo(else: a))) == "foo(else: a)"
+      assert quoted_to_string(quote(do: foo(catch: a))) == "foo(catch: a)"
+    end
+  end
+
+  describe "quoted_to_algebra/2 escapes" do
+    test "strings with slash escapes" do
+      assert quoted_to_string(quote(do: "\a\b\d\e\f\n\r\t\v"), escape: false) ==
+               ~s/"\a\b\d\e\f\n\r\t\v"/
+
+      assert quoted_to_string(quote(do: "\a\b\d\e\f\n\r\t\v")) ==
+               ~s/"\\a\\b\\d\\e\\f\\n\\r\\t\\v"/
+    end
+
+    test "strings with non printable characters" do
+      assert quoted_to_string(quote(do: "\x00\x01\x10"), escape: false) == ~s/"\x00\x01\x10"/
+      assert quoted_to_string(quote(do: "\x00\x01\x10")) == ~s/"<<0, 1, 16>>"/
+    end
+
+    test "charlists with slash escapes" do
+      assert quoted_to_string(quote(do: '\a\b\e\n\r\t\v'), escape: false) ==
+               ~s/'\a\b\e\n\r\t\v'/
+
+      assert quoted_to_string(quote(do: '\a\b\e\n\r\t\v')) ==
+               ~s/'\\a\\b\\e\\n\\r\\t\\v'/
+    end
+
+    test "charlists with non printable characters" do
+      assert quoted_to_string(quote(do: '\x00\x01\x10'), escape: false) == ~S/[0, 1, 16]/
+      assert quoted_to_string(quote(do: '\x00\x01\x10')) == ~S/[0, 1, 16]/
+    end
+
+    test "atoms" do
+      assert quoted_to_string(quote(do: :"a\nb\tc"), escape: false) == ~s/:"a\nb\tc"/
+      assert quoted_to_string(quote(do: :"a\nb\tc")) == ~S/:"a\nb\tc"/
+    end
+
+    test "atoms with interpolations" do
+      assert quoted_to_string(quote(do: :"foo\n#{bar}\tbaz"), escape: false) ==
+               ~s[:"foo\n\#{bar}\tbaz"]
+
+      assert quoted_to_string(quote(do: :"foo\n#{bar}\tbaz")) == ~S[:"foo\n#{bar}\tbaz"]
+
+      assert quoted_to_string(quote(do: :"foo\"bar"), escape: false) == ~s[:"foo\\"bar"]
+      assert quoted_to_string(quote(do: :"foo\"bar")) == ~S[:"foo\\"bar"]
+
+      assert quoted_to_string(quote(do: :"one\n\"#{2}\"\nthree"), escape: false) ==
+               ~s[:"one\n\\"\#{2}\\"\nthree"]
+
+      assert quoted_to_string(quote(do: :"one\n\"#{2}\"\nthree")) == ~S[:"one\n\\"#{2}\\"\nthree"]
```

Comment:
I have a doubt regarding the escaping/unescaping of atom interpolations. It seems that some characters like `\n` are escaped correctly, but others like `\"` get the `\` escaped twice.
In these lines I had to put double backward slashes on the right hand sides because of this, but while it is the same escaping that `Macro.to_string` does, it's not exactly the same behavior as what `:unescape` option of `Code.string_to_quoted` produces.
Is this alright?

Reply:
The above is definitely wrong. This:

```elixir
:"one\n\\"#{2}\\"\nthree"
```

Is actually the same as:

```elixir
:"one\n\\"       #{2}\\"\nthree"
```

So the #{... is being treated as a comment. The one with `escape: false` is also wrong for the same reason.

Reply:
I fixed the test so it expects the correct output. It is now failing as it should.

I will try to fix this tomorrow.

Reply:
I think it its fixed now c: